tests/model_fields/models.py

```diff
@@ -609,3 +609,75 @@ class GeneratedModelNullVirtual(models.Model):
 
     class Meta:
         required_db_features = {"supports_virtual_generated_columns"}
+
+
+class GeneratedModelUniqueConstraint(models.Model):
+    name = models.CharField(max_length=10)
+    lower_name = models.GeneratedField(
+        expression=Lower("name"),
+        output_field=models.CharField(max_length=10),
+        db_persist=True,
+    )
+
+    class Meta:
+        required_db_features = {"supports_stored_generated_columns"}
+        constraints = [
+            models.UniqueConstraint(Lower("name"), name="name_uniq_generated")
```

Comment:
```suggestion
            models.UniqueConstraint(fields=["name"], name="name_uniq_generated"),
            models.UniqueConstraint(fields=["lower_name"], name="name_lower_uniq"),
```
If I update to have a unique constraint on `lower_name`, `StoredGeneratedFieldTests.test_full_clean_with_unique_constraint` still fails with an attribute error `AttributeError: Cannot read a generated field from an unsaved model.`

On your earlier question which was roughly around whether `GeneratedField`s fields should be ignored from `Model.validate_constraints(exclude=None)` - I _think_ so
Agree we should document it maybe as an admonition to `Model.validate_constraints(exclude=None)` (we also need a release note for 5.0.7 for this change).
Not allowing to create constraints on `GeneratedField`s seems harsh to me when it "works"

Reply:
Hi @sarahboyce, thanks for the review. I'll take a look into why using the `GeneratedField` in `UniqueConstraint.fields` still fails and suggest a change.

> Not allowing to create constraints on `GeneratedFields` seems harsh to me when it "works"
Yes other than `.full_clean()` they appear to work correctly. I'll investigate what would be required to make `full_clean()` work correctly (rather than forcing the exclusion of generated fields from constraints) and suggest an alternative solution.

Reply:
I have added the tests as you suggested, and applied a change to `Constraint.validate()` which was also causing the `GeneratedField` attribute to be accessed before a model instance had been saved. This was because constraints initialised as  `UniqueConstraint(fields=[], ...)` are hanlded differently to constraints initialised via `UniqueConstraint(<expression>, ...)`.

I'll add my thoughts on the bigger picture in a comment  on the main PR.